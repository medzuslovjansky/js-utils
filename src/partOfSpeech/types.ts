export type PartOfSpeech =
  | Adjective
  | Adverb
  | Conjunction
  | Interjection
  | Noun
  | Numeral
  | Participle
  | Particle
  | Phrase
  | Preposition
  | Prefix
  | Pronoun
  | Suffix
  | Verb;

export type Adjective = {
  name: 'adjective';
  positive: boolean;
  comparative: boolean;
  superlative: boolean;
  absolute: boolean;
  possessive: boolean;
};

export type Adverb = {
  name: 'adverb';
  positive: boolean;
  comparative: boolean;
  superlative: boolean;
  absolute: boolean;
};

export type Conjunction = {
  name: 'conjunction';
};

export type Interjection = {
  name: 'interjection';
};

export type Noun = {
  name: 'noun';
  // TODO: change to gender?: - for parsing errors
  gender: 'masculine' | 'feminine' | 'neuter' | 'masculineOrFeminine';
  masculine: boolean;
  animate: boolean;
  feminine: boolean;
  neuter: boolean;
  singular: boolean;
  plural: boolean;
  indeclinable: boolean;
};

export type Numeral = {
  name: 'numeral';
  // TODO: change to type?: - for parsing errors
  type:
    | 'cardinal'
    | 'collective'
    | 'differential'
    | 'fractional'
    | 'multiplicative'
    | 'ordinal'
    | 'substantivized';
  cardinal: boolean;
  collective: boolean;
  differential: boolean;
  fractional: boolean;
  multiplicative: boolean;
  ordinal: boolean;
  substantivized: boolean;
};

export type Participle = {
  name: 'participle';
  active: boolean;
  passive: boolean;
  present: boolean;
  past: boolean;
};

export type Particle = {
  name: 'particle';
};

export type Phrase = {
  name: 'phrase';
};

export type Prefix = {
  name: 'prefix';
};

export type Preposition = {
  name: 'preposition';
};

export type Pronoun = {
  name: 'pronoun';
  // TODO: change to type?: - for parsing errors
  type:
    | 'demonstrative'
    | 'indefinite'
    | 'interrogative'
    | 'personal'
    | 'possessive'
    | 'reciprocal'
    | 'reflexive'
    | 'relative';
  demonstrative: boolean;
  indefinite: boolean;
  interrogative: boolean;
  personal: boolean;
  possessive: boolean;
  reciprocal: boolean;
  reflexive: boolean;
  relative: boolean;
};

export type Suffix = {
  name: 'suffix';
};

export type Verb = {
  name: 'verb';
  auxiliary: boolean;
  imperfective: boolean;
  intransitive: boolean;
  perfective: boolean;
  reflexive: boolean;
  transitive: boolean;
};
