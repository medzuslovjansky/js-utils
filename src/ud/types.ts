export interface Word {
  form: string;
  lemma: string;
  uPosTag: string;
  xPosTag?: string;
  feats: object;
}
