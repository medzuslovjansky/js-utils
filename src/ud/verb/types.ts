import type { Word } from '../types';

export interface Aux extends Word {
  uPosTag: 'AUX';
}

export interface Verb extends Word {
  uPosTag: 'VERB';
  feats: VerbFeatures;
}

export interface VerbFeatures {
  aspect?: 'Imp' | 'Perf';
  gender?: 'Masc' | 'Fem' | 'Neut';
  mood?: 'Ind' | 'Cnd' | 'Imp' | 'Sub';
  number?: 'Sing' | 'Plur';
  person?: '1' | '2' | '3';
  tense?: 'Past' | 'Pres' | 'Fut' | 'Imp';
  verbForm?: 'Fin' | 'Inf' | 'Part' | 'Conv';
  voice?: 'Act' | 'Pass';
}
