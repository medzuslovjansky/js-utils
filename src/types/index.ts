export * from './pos';
export * from './intelligibility';
export * from '../core';
