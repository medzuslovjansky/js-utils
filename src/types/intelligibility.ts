export enum SlavicLanguage {
  Belarusian = 'be',
  Bosnian = 'bs',
  Bulgarian = 'bg',
  Croatian = 'hr',
  Czech = 'cs',
  Kashubian = 'csb',
  LowerSorbian = 'dsb',
  Macedonian = 'mk',
  Montenegrin = 'cnr',
  ChurchSlavonic = 'cu',
  OldRussian = 'orv',
  Polabian = 'pox',
  Polish = 'pl',
  Russian = 'ru',
  Rusyn = 'rue',
  Serbian = 'sr',
  Silesian = 'szl',
  Slovak = 'sk',
  Slovenian = 'sl',
  Ukrainian = 'uk',
  UpperSorbian = 'hsb',
}

export type SlavicRegionalTag =
  | 'z'
  | 'v'
  | 'j'
  | 'be'
  | 'bg'
  | 'bm'
  | 'cs'
  | 'csb'
  | 'cu'
  | 'cz'
  | 'dsb'
  | 'hr'
  | 'hsb'
  | 'iw'
  | 'mk'
  | 'ns'
  | 'ocs'
  | 'pl'
  | 'ps'
  | 're'
  | 'ru'
  | 'rue'
  | 'sb'
  | 'sh'
  | 'sk'
  | 'sl'
  | 'sr'
  | 'ub'
  | 'uk'
  | 'yu'
  | 'i'
  | 'ij'
  | 'in'
  | 'iz'
  | 'jc'
  | 'jn'
  | 'n'
  | 'sx';

export enum VoteStatus {
  Common = 1,
  Regional = 2,
  Individual = 3,
  Archaic = 4,
  Artificial = 5,
  Questionable = 9,
  Deprecated = 99,
}

export enum Genesis {
  Arabic = 'A',
  Deutsch = 'D',
  English = 'E',
  French = 'F',
  German = 'G',
  International = 'I',
  Artificial = 'M',
  Netherlands = 'N',
  Oriental = 'O',
  Slavic = 'S',
  Turkic = 'T',
}

export enum CrudeIntelligibilityLevel {
  Full = 1,
  Incomplete = 0.5,
  Disputed = 0.5,
  Other1 = 0.5,
  Unknown = 0.1,
}

export type CrudeIntelligibilityReport = Partial<
  Record<SlavicLanguage, CrudeIntelligibilityLevel>
>;
