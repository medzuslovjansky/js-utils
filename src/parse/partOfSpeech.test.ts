import parsePartOfSpeech from './partOfSpeech';
import { PartOfSpeech } from '../types';

type ParseTestTuple = [input: string, expected: Partial<PartOfSpeech>];

describe('parsePartOfSpeech', () => {
  test.each([
    ['adj.', { name: 'adjective' }],
    ['adv.', { name: 'adverb' }],
    ['conj.', { name: 'conjunction' }],
    ['f.', { name: 'noun', feminine: true }],
    ['f.indecl.', { name: 'noun', feminine: true, indeclinable: true }],
    ['f.pl.', { name: 'noun', feminine: true, plural: true }],
    ['f.sg.', { name: 'noun', feminine: true, singular: true }],
    ['intj.', { name: 'interjection' }],
    ['m.', { name: 'noun', masculine: true }],
    ['m./f.', { name: 'noun', masculine: true, feminine: true }],
    ['m.anim.', { name: 'noun', masculine: true, animate: true }],
    ['m.indecl.', { name: 'noun', masculine: true, indeclinable: true }],
    ['m.pl.', { name: 'noun', masculine: true, plural: true }],
    ['m.sg.', { name: 'noun', masculine: true, singular: true }],
    ['n.', { name: 'noun', neuter: true }],
    ['n.indecl.', { name: 'noun', neuter: true, indeclinable: true }],
    ['n.pl.', { name: 'noun', neuter: true, plural: true }],
    ['n.sg.', { name: 'noun', neuter: true, singular: true }],
    ['noun indecl.', { name: 'noun', indeclinable: true }],
    ['num.', { name: 'numeral' }],
    ['num.card.', { name: 'numeral', cardinal: true }],
    ['num.coll.', { name: 'numeral', collective: true }],
    ['num.diff.', { name: 'numeral', differential: true }],
    ['num.fract.', { name: 'numeral', fractional: true }],
    ['num.mult.', { name: 'numeral', multiplicative: true }],
    ['num.ord.', { name: 'numeral', ordinal: true }],
    ['num.subst.', { name: 'numeral', substantivized: true }],
    ['particle', { name: 'particle' }],
    ['phrase', { name: 'phrase' }],
    ['prefix', { name: 'prefix' }],
    ['prep.', { name: 'preposition' }],
    ['pron.dem.', { name: 'pronoun', demonstrative: true }],
    ['pron.indef.', { name: 'pronoun', indefinite: true }],
    ['pron.int.', { name: 'pronoun', intensive: true }],
    ['pron.pers.', { name: 'pronoun', personal: true }],
    ['pron.poss.', { name: 'pronoun', possessive: true }],
    ['pron.rec.', { name: 'pronoun', reciprocal: true }],
    ['pron.refl.', { name: 'pronoun', reflexive: true }],
    ['pron.rel.', { name: 'pronoun', relative: true }],
    ['suffix', { name: 'suffix' }],
    ['v. ipf.', { name: 'verb', imperfective: true }],
    ['v.aux. ipf.', { name: 'verb', auxiliary: true, imperfective: true }],
    ['v.aux. pf.', { name: 'verb', auxiliary: true, perfective: true }],
    ['v.intr. ipf.', { name: 'verb', intransitive: true, imperfective: true }],
    [
      'v.intr. ipf./pf.',
      {
        name: 'verb',
        intransitive: true,
        imperfective: true,
        perfective: true,
      },
    ],
    ['v.intr. pf.', { name: 'verb', intransitive: true, perfective: true }],
    ['v.ipf.', { name: 'verb', imperfective: true }],
    ['v.pf.', { name: 'verb', perfective: true }],
    ['v.refl. ipf.', { name: 'verb', reflexive: true, imperfective: true }],
    [
      'v.refl. ipf./pf.',
      { name: 'verb', reflexive: true, imperfective: true, perfective: true },
    ],
    ['v.refl. pf.', { name: 'verb', reflexive: true, perfective: true }],
    ['v.tr. ipf.', { name: 'verb', transitive: true, imperfective: true }],
    [
      'v.tr. ipf./pf.',
      { name: 'verb', transitive: true, imperfective: true, perfective: true },
    ],
    ['v.tr. pf.', { name: 'verb', transitive: true, perfective: true }],
    ['v.tr.ipf', { name: 'verb', transitive: true, imperfective: true }],
    ['#v.tr.ipf', { name: 'verb', transitive: true, imperfective: true }],
  ] as ParseTestTuple[])(
    'should expand %j to object',
    (abbr: ParseTestTuple[0], expected: ParseTestTuple[1]) => {
      const actual = parsePartOfSpeech(abbr);
      if (!actual) {
        fail('should have returned non-null');
      }

      const pattern: any = { ...expected };
      for (const aKey of Object.keys(actual)) {
        const key: keyof PartOfSpeech = aKey as any;
        if (typeof actual[key] === 'boolean' && !(key in pattern)) {
          pattern[key] = false;
        }
      }

      expect(actual).toMatchObject(pattern);
    },
  );

  it('should throw for non-strings', () => {
    expect(() => parsePartOfSpeech(null as any)).toThrowErrorMatchingSnapshot();
  });

  it('should throw for unknown abbreviations', () => {
    expect(() => parsePartOfSpeech('unknown')).toThrowErrorMatchingSnapshot();
  });
});
