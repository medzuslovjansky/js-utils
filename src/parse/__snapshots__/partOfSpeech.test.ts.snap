// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePartOfSpeech should throw for non-strings 1`] = `"Expected a string, but got: null"`;

exports[`parsePartOfSpeech should throw for unknown abbreviations 1`] = `"Failed to parse the part of speech in: unknown"`;
